catroid/src/main/java/org/catrobat/catroid/web/WebConnection.java

```diff
@@ -0,0 +1,140 @@
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class WebConnection {
+	private final OkHttpClient okHttpClient;
+	private static final String USER_AGENT = "Mozilla/5.0 (compatible; Catrobatbot/1.0; +https://catrob.at/bot)";
+
+	private Call call;
+	private String url;
+	private WeakReference<WebRequestListener> weakListenerReference;
+
+	public WebConnection() {
+		okHttpClient = new OkHttpClient();
+		ConnectionSpec modernTlsSpec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS).build();
+		ConnectionSpec compatibleTlsSpec = new ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS).build();
+		ConnectionSpec cleartextSpec = new ConnectionSpec.Builder(ConnectionSpec.CLEARTEXT).build();
+		okHttpClient.setConnectionSpecs(ImmutableList.of(modernTlsSpec, compatibleTlsSpec, cleartextSpec));
+	}
+
+	public WebConnection(OkHttpClient okHttpClient) {
+		this.okHttpClient = okHttpClient;
+	}
+
+	public void setListener(WebRequestListener listener) {
+		this.weakListenerReference = new WeakReference<>(listener);
+	}
+
+	public void setUrl(String url) {
+		this.url = url;
+	}
+
+	private synchronized WebRequestListener getListenerAndSetReferenceToNull() {
+		if (this.weakListenerReference != null) {
+			WebRequestListener listener = this.weakListenerReference.get();
+			this.weakListenerReference = null;
+			return listener;
+		}
+		return null;
+	}
+
+	public synchronized boolean sendWebRequest() {
+		try {
+			checkIfUrlIsValid();
+		} catch (RuntimeException e) {
+			return false;
+		}
+
+		Request request = new Request.Builder()
+				.url(url)
+				.header("User-Agent", USER_AGENT)
+				.build();
+
+		call = okHttpClient.newCall(request);
+		call.enqueue(getCallback());
+		return true;
+	}
+
+	private Callback getCallback() {
+		return new Callback() {
+			@Override
+			public void onFailure(Request request, IOException e) {
+			}
+
+			@Override
+			public void onResponse(Response response) throws IOException {
+				WebRequestListener listener = getListenerAndSetReferenceToNull();
+				if (listener != null) {
+					if (response.isSuccessful()) {
+						listener.onRequestFinished(response.body().string());
+					} else {
+						listener.onRequestFinished(String.valueOf(response.code()));
+					}
+				}
+			}
+		};
+	}
+
+	public void cancelCall() {
+		if (call != null) {
+			okHttpClient.getDispatcher().getExecutorService().execute(new Runnable() {
+				@Override
+				public void run() {
+					WebRequestListener listener = getListenerAndSetReferenceToNull();
+					if (listener != null) {
+						call.cancel();
+						listener.onCancelledCall();
+					}
+				}
+			});
+		}
+	}
+
+	private void checkIfUrlIsValid() {
+		try {
+			new URL(url);
+		} catch (MalformedURLException e) {
+			throw new RuntimeException("Malformed URL: " + url, e);
```

Comment:
why the heck a RuntimeException?
dont write check() methods that dont have a return value and throw a runtime exception instead that quite possibly nobody knows that its throwing (because its a runtimeException you dont have to add it to the signature and nobody else will be warned about this by the IDE)

either remove the method and do the try catch block on the malformedURLExecption in sendWebRequest, or have this return true or false. (and if so, rename to something meaningful like "urlValid()" and add the url to the parameter and not operate on the member directly.